storage: check for additional split work after a split

petermattis opened the pull request:
After splitting a range, further split work might be required if the
zone config changed or a table was created concurrently or if the range
was much larger than the target size.

Before this change, TestSplitAtTableBoundary had a wide variance in how
long it would take. Sometimes it would completely in less than a second
while other times it would take 10-20 seconds. With this change it
reliably completes in less than a second.

Fixes #10160

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10232)

<!-- Reviewable:end -->

petermattis committed:
storage: check for additional split work after a split

After splitting a range, further split work might be required if the
zone config changed or a table was created concurrently or if the range
was much larger than the target size.

Before this change, TestSplitAtTableBoundary had a wide variance in how
long it would take. Sometimes it would completely in less than a second
while other times it would take 10-20 seconds. With this change it
reliably completes in less than a second.

Fixes #10160

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-:-KV0UupmTa_zs6EflxIw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/table_split_test.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0gERTJBydPelK9BZc:-KV0gERTJBydPelK9BZd:b4qxw4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L43)):*
> ```Go
> 	defer tc.Stopper().Stop()
> 	if err := tc.WaitForFullReplication(); err != nil {
> 		t.Error(err)
> ```

should be Fatal

---

*[pkg/sql/table_split_test.go, line 47 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0gFuMtTeFG-p9HL9B:-KV0gFuMtTeFG-p9HL9C:bbknbra) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L47)):*
> ```Go
> 
> 	sqlDB := sqlutils.MakeSQLRunner(t, tc.Conns[0])
> 	_ = sqlDB.Exec(`CREATE DATABASE test`)
> ```

should check these errors.

---

*[pkg/sql/table_split_test.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0h34x4V3nS2pESWG1:-KV0h34x4V3nS2pESWG2:b9asq2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L50)):*
> ```Go
> 	_ = sqlDB.Exec(`CREATE TABLE test.t (k SERIAL PRIMARY KEY, v INT)`)
> 
> 	tableIDQuery := `
> ```

optional: const

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-:-KV0h6VRmOAOXfCY0L8H:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/table_split_test.go, line 43 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0gERTJBydPelK9BZc:-KV0k80DFTcFDVsWiPq-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should be Fatal
</details>
Done.

---

*[pkg/sql/table_split_test.go, line 47 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0gFuMtTeFG-p9HL9B:-KV0kAk6Lu5jiHcR_0xW:bczb86g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L47)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should check these errors.
</details>
Renamed `sqlDB` to `runner` to make this clear that this isn't a `database/sql.DB`. Also removed the assignment to `_` which wasn't necessary.

---

*[pkg/sql/table_split_test.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232#-KV0h34x4V3nS2pESWG1:-KV0kHqGJ7vLufCXkXda:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622a428868e63a6af4c42df0a6b8baafbb84b85/pkg/sql/table_split_test.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> optional: const
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10232)*
<!-- Sent from Reviewable.io -->
